## Preprocessing

In [2]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [3]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(columns=['EIN', 'STATUS'])
application_df.head()

NAME APPLICATION_TYPE  \
0              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION     INCOME_AMT  \
0       Independent          C1000    ProductDev   Association              0   
1       Independent          C2000  Preservation  Co-operative         1-9999   
2  CompanySponsored          C3000    ProductDev   Association              0   
3  CompanySponsored          C2000  Preservation         Trust    10000-24999   
4       Independent          C1000     Heathcare         Trust  100000-499999   

  SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0                      N     5000              1  
1                      N   108590              1  
2                      N     5000              0  
3                      N     6692              1  
4                      N   142590              1

In [4]:
application_df.dtypes

NAME                      object
APPLICATION_TYPE          object
AFFILIATION               object
CLASSIFICATION            object
USE_CASE                  object
ORGANIZATION              object
INCOME_AMT                object
SPECIAL_CONSIDERATIONS    object
ASK_AMT                    int64
IS_SUCCESSFUL              int64
dtype: object

In [5]:
# Determine the number of unique values in each column.
unique_values = application_df.nunique()
print(unique_values)

NAME                      19568
APPLICATION_TYPE             17
AFFILIATION                   6
CLASSIFICATION               71
USE_CASE                      5
ORGANIZATION                  4
INCOME_AMT                    9
SPECIAL_CONSIDERATIONS        2
ASK_AMT                    8747
IS_SUCCESSFUL                 2
dtype: int64


In [6]:
# Look at APPLICATION_TYPE value counts for binning
apptype_column = application_df['APPLICATION_TYPE'].value_counts()
print(apptype_column)

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64


In [7]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
other = 400
application_types_to_replace = application_df['APPLICATION_TYPE'].value_counts()[application_df['APPLICATION_TYPE'].value_counts() < other].index.tolist()

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [8]:
# Look at CLASSIFICATION value counts for binning
classification_column = application_df['CLASSIFICATION'].value_counts()
print(classification_column)

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64


In [9]:
# You may find it helpful to look at CLASSIFICATION value counts >1
print(classification_column[classification_column>1])

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64


In [10]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
other = 2
classifications_to_replace = application_df['CLASSIFICATION'].value_counts()[application_df['CLASSIFICATION'].value_counts() < other].index.tolist()


# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
Other       26
C8000       20
C7120       18
C1500       16
C6000       15
C1800       15
C1250       14
C8200       11
C1238       10
C1278       10
C1237        9
C1235        9
C7210        7
C1720        6
C2400        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C1267        2
C1256        2
C1234        2
C1246        2
C3200        2
Name: CLASSIFICATION, dtype: int64

In [11]:
# Look at NAME counts for binning
name_column = application_df['NAME'].value_counts()
print(name_column)

PARENT BOOSTER USA INC                                                  1260
TOPS CLUB INC                                                            765
UNITED STATES BOWLING CONGRESS INC                                       700
WASHINGTON STATE UNIVERSITY                                              492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                          408
                                                                        ... 
ST LOUIS SLAM WOMENS FOOTBALL                                              1
AIESEC ALUMNI IBEROAMERICA CORP                                            1
WEALLBLEEDRED ORG INC                                                      1
AMERICAN SOCIETY FOR STANDARDS IN MEDIUMSHIP & PSYCHICAL INVESTIGATI       1
WATERHOUSE CHARITABLE TR                                                   1
Name: NAME, Length: 19568, dtype: int64


In [13]:
# Choose a cutoff value and create a list of names to be replaced
# use the variable name `names_to_replace`
other = 10
names_to_replace = application_df['NAME'].value_counts()[application_df['NAME'].value_counts() < other].index.tolist()


# Replace in dataframe
for cls in names_to_replace:
    application_df['NAME'] = application_df['NAME'].replace(cls,"Other")

# Check to make sure binning was successful
application_df['NAME'].value_counts()

Other                                      21022
PARENT BOOSTER USA INC                      1260
TOPS CLUB INC                                765
UNITED STATES BOWLING CONGRESS INC           700
WASHINGTON STATE UNIVERSITY                  492
                                           ...  
CASCADE 4-H FOUNDATION                        10
FREE & ACCEPTED MASONS OF WASHINGTON          10
NEW MEXICO GARDEN CLUBS INC                   10
NATIONAL ASSOCIATION OF HISPANIC NURSES       10
UNION OF CALIFORNIA STATE WORKERS             10
Name: NAME, Length: 223, dtype: int64

In [16]:
unique_values_new = application_df.nunique()
print(unique_values_new)

NAME                       223
APPLICATION_TYPE             9
AFFILIATION                  6
CLASSIFICATION              46
USE_CASE                     5
ORGANIZATION                 4
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64


In [17]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df = pd.get_dummies(application_df)
application_df.head()

ASK_AMT  IS_SUCCESSFUL  NAME_ACE MENTOR PROGRAM OF AMERICA INC  \
0     5000              1                                       0   
1   108590              1                                       0   
2     5000              0                                       0   
3     6692              1                                       0   
4   142590              1                                       0   

   NAME_AIR FORCE ASSOCIATION  NAME_ALABAMA FEDERATION OF WOMENS CLUBS  \
0                           0                                        0   
1                           0                                        0   
2                           0                                        0   
3                           0                                        0   
4                           0                                        0   

   NAME_ALABAMA TREASURE FOREST ASSOCIATION  NAME_ALPHA PHI SIGMA  \
0                                         0                     0   
1                                         0                     0   
2                                         0                     0   
3                                         0                     0   
4                                         0                     0   

   NAME_ALPHA PHI SIGMA INC  NAME_ALTRUSA INTERNATIONAL FOUNDATION INC  \
0                         0                                          0   
1                         0                                          0   
2                         0                                          0   
3                         0                                          0   
4                         0                                          0   

   NAME_AMATEUR ATHLETIC UNION OF THE UNITED STATES  ...  INCOME_AMT_1-9999  \
0                                                 0  ...                  0   
1                                                 0  ...                  1   
2                                                 0  ...                  0   
3                                                 0  ...                  0   
4                                                 0  ...                  0   

   INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
0                       0                         0                   0   
1                       0                         0                   0   
2                       0                         0                   0   
3                       1                         0                   0   
4                       0                         1                   0   

   INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  INCOME_AMT_50M+  \
0                 0                       0                0   
1                 0                       0                0   
2                 0                       0                0   
3                 0                       0                0   
4                 0                       0                0   

   INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  
0                  0                         1                         0  
1                  0                         1                         0  
2                  0                         1                         0  
3                  0                         1                         0  
4                  0                         1                         0  

[5 rows x 306 columns]

In [18]:
# Split our preprocessed data into our features and target arrays
y = application_df['IS_SUCCESSFUL']
X = application_df.drop(columns='IS_SUCCESSFUL')

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [19]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [20]:
number_input_features = len( X_train_scaled[0])
print(number_input_features)

305


In [21]:
# Determine how many input features there are
num_input_features = application_df.shape[1] - 1
print("Number of input features:", num_input_features)

Number of input features: 305


## First Optimization Attempt
### For this 1st optimization attempt I reduced the cutoff point for the APPLICATION_TYPE column from 528 to 400 creating fewer bins to look at. I also reduced the cutoff point for the CLASSIFICATION column from 1883 to 10 to increase the number of data points we looked at. I also tried several different feature combinations and eventually landed on removing EIN and STATUS. After adding NAME back I saw a large increase in results so I added that back and binned the groups to only look at NAMEs with more than 10 occurances as my colab was crashing if I tried to do all the names at once. I also used the same number of neurons as input features.

In [30]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
input_features = application_df.shape[1] - 1

neurons = input_features


nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=neurons, activation="relu", input_dim=input_features))


# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_6 (Dense)             (None, 305)               93330     
                                                                 
 dense_7 (Dense)             (None, 1)                 306       
                                                                 
Total params: 93636 (365.77 KB)
Trainable params: 93636 (365.77 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [31]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [32]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 3s 3ms/step - loss: 0.4805 - accuracy: 0.7717
Epoch 2/100
804/804 [==============================] - 3s 3ms/step - loss: 0.4502 - accuracy: 0.7807
Epoch 3/100
804/804 [==============================] - 3s 3ms/step - loss: 0.4441 - accuracy: 0.7836
Epoch 4/100
804/804 [==============================] - 4s 5ms/step - loss: 0.4407 - accuracy: 0.7842
Epoch 5/100
804/804 [==============================] - 3s 3ms/step - loss: 0.4398 - accuracy: 0.7873
Epoch 6/100
804/804 [==============================] - 3s 3ms/step - loss: 0.4393 - accuracy: 0.7871
Epoch 7/100
804/804 [==============================] - 3s 4ms/step - loss: 0.4372 - accuracy: 0.7869
Epoch 8/100
804/804 [==============================] - 4s 5ms/step - loss: 0.4369 - accuracy: 0.7881
Epoch 9/100
804/804 [==============================] - 3s 4ms/step - loss: 0.4352 - accuracy: 0.7882
Epoch 10/100
804/804 [==============================] - 4s 5ms/step - loss: 0.4351 - accura

In [33]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5541 - accuracy: 0.7752 - 563ms/epoch - 2ms/step
Loss: 0.5541175007820129, Accuracy: 0.7751603722572327


## Second Optimization Attempt
### For the 2nd optimization attempt I tried doubling the number of neurons I was using in the model to see if that would have an impact on the accuracy score. I also added a hidden layer to see how that might impact the code.

In [34]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
input_features = application_df.shape[1] - 1
# In class we learned we should use 2-3 times the # of input features for neurons for an initial model

# Same neuron number as that seemed to ever so slightly improve it
neurons1 = input_features*2

nn2 = tf.keras.models.Sequential()

# First hidden layer
nn2.add(tf.keras.layers.Dense(units=neurons1, activation="relu", input_dim=input_features))

# Second Hidden Layer
nn2.add(tf.keras.layers.Dense(units=neurons1, activation="relu"))


# Output layer
nn2.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn2.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_8 (Dense)             (None, 610)               186660    
                                                                 
 dense_9 (Dense)             (None, 610)               372710    
                                                                 
 dense_10 (Dense)            (None, 1)                 611       
                                                                 
Total params: 559981 (2.14 MB)
Trainable params: 559981 (2.14 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [35]:
# Compile the model
nn2.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [36]:
# Train the model
fit_model = nn2.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 9s 10ms/step - loss: 0.4754 - accuracy: 0.7729
Epoch 2/100
804/804 [==============================] - 7s 9ms/step - loss: 0.4443 - accuracy: 0.7835
Epoch 3/100
804/804 [==============================] - 9s 11ms/step - loss: 0.4407 - accuracy: 0.7865
Epoch 4/100
804/804 [==============================] - 8s 11ms/step - loss: 0.4391 - accuracy: 0.7877
Epoch 5/100
804/804 [==============================] - 7s 9ms/step - loss: 0.4367 - accuracy: 0.7880
Epoch 6/100
804/804 [==============================] - 9s 11ms/step - loss: 0.4366 - accuracy: 0.7879
Epoch 7/100
804/804 [==============================] - 8s 10ms/step - loss: 0.4350 - accuracy: 0.7886
Epoch 8/100
804/804 [==============================] - 8s 10ms/step - loss: 0.4335 - accuracy: 0.7892
Epoch 9/100
804/804 [==============================] - 9s 11ms/step - loss: 0.4326 - accuracy: 0.7900
Epoch 10/100
804/804 [==============================] - 7s 9ms/step - loss: 0.4314 -

In [37]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn2.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 1.0416 - accuracy: 0.7796 - 882ms/epoch - 3ms/step
Loss: 1.0416241884231567, Accuracy: 0.7795918583869934


## Third Optimization Attempt
### For the third Optimization Attempt I created a method that allowed kerastuner to decide the optimal activation function, number of neurons, and hidden layers.

In [45]:
# Create a method that creates a new Sequential model with hyperparameter options
def create_model(hp):

    input_features = application_df.shape[1] - 1
    nn3 = tf.keras.models.Sequential()

    # Allow kerastuner to decide which activation function to use in hidden layers
    activation = hp.Choice('activation',['relu','tanh'])

    # Allow kerastuner to decide number of neurons in first layer
    nn3.add(tf.keras.layers.Dense(units=hp.Int('first_units',
        min_value=1,
        max_value=400,
        step=5), activation=activation, input_dim=input_features))

    # Allow kerastuner to decide number of hidden layers and neurons in hidden layers
    for i in range(hp.Int('num_layers', 1, 5)):
        nn3.add(tf.keras.layers.Dense(units=hp.Int('units_' + str(i),
            min_value=1,
            max_value=400,
            step=5),
            activation=activation))

    nn3.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

    # Compile the model
    nn3.compile(loss="binary_crossentropy", optimizer='adam', metrics=["accuracy"])

    return nn3

In [46]:
pip install keras-tuner

In [47]:
# Import the kerastuner library
import keras_tuner as kt

tuner = kt.Hyperband(
    create_model,
    objective="val_accuracy",
    max_epochs=20,
    hyperband_iterations=2)

Reloading Tuner from ./untitled_project/tuner0.json


In [48]:
# Run the kerastuner search for best hyperparameters
tuner.search(X_train_scaled,y_train,epochs=20,validation_data=(X_test_scaled,y_test))

Trial 56 Complete [00h 00m 50s]
val_accuracy: 0.7344606518745422

Best val_accuracy So Far: 0.783790111541748
Total elapsed time: 00h 27m 21s


In [49]:
# Search for the best hyperparameters
best_hyper = tuner.get_best_hyperparameters(1)[0]
best_hyper.values

{'activation': 'relu',
 'first_units': 16,
 'num_layers': 5,
 'units_0': 11,
 'units_1': 26,
 'units_2': 16,
 'units_3': 151,
 'units_4': 16,
 'tuner/epochs': 20,
 'tuner/initial_epoch': 7,
 'tuner/bracket': 2,
 'tuner/round': 2,
 'tuner/trial_id': '0039'}

In [50]:
# Evaluate best model against full test data
best_model = tuner.get_best_models(1)[0]
model_loss, model_accuracy = best_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.4568 - accuracy: 0.7838 - 1s/epoch - 4ms/step
Loss: 0.45676347613334656, Accuracy: 0.783790111541748


In [52]:
# Export our model to HDF5 file
from tensorflow.keras.models import save_model
from google.colab import drive
drive.mount('/content/drive')
nn.save("/content/drive/My Drive/Colab Notebooks/AlphabetSoupCharity_Optimization.h5")

Mounted at /content/drive


/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(
